File: DID.Native.SDK/src/issuer.c
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <assert.h>

#include "ela_did.h"
#include "diderror.h"
#include "did.h"
#include "credential.h"
#include "crypto.h"
#include "issuer.h"
#include "didstore.h"
#include "diddocument.h"

extern const char *ProofType;

Issuer *Issuer_Create(DID *did, DIDURL *signkey, DIDStore *store)
{
    Issuer *issuer;
    DIDDocument *doc;
    bool isAuthKey;

    if (!did || !store) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    //doc = DID_Resolve(did);
    doc = DIDStore_LoadDID(store, did);
    if (!doc)
        return NULL;

    if (!signkey)
        signkey = DIDDocument_GetDefaultPublicKey(doc);
    else {
        isAuthKey = DIDDocument_IsAuthenticationKey(doc, signkey);
        if (!isAuthKey) {
            DIDError_Set(DIDERR_INVALID_KEY, "Invalid authentication key.");
            DIDDocument_Destroy(doc);
            return NULL;
        }
    }

    if (!DIDStore_ContainsPrivateKey(store, did, signkey)) {
        DIDError_Set(DIDERR_DIDSTORE_ERROR, "Missing private key paired with signkey.");
        DIDDocument_Destroy(doc);
        return NULL;
    }

    issuer = (Issuer*)calloc(1, sizeof(Issuer));
    if (!issuer) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Malloc buffer for issuer failed.");
        DIDDocument_Destroy(doc);
        return NULL;
    }

    issuer->signer = doc;
    DIDURL_Copy(&issuer->signkey, signkey);
    return issuer;
}

void Issuer_Destroy(Issuer *issuer)
{
    if (!issuer)
        return;

    if (issuer->signer)
        DIDDocument_Destroy(issuer->signer);

    free(issuer);
}

DID *Issuer_GetSigner(Issuer *issuer)
{
    if (!issuer) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    return &issuer->signer->did;
}

DIDURL *Issuer_GetSignKey(Issuer *issuer)
{
    if (!issuer) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    return &issuer->signkey;
}

static Credential *issuer_generate_credential(Issuer *issuer, DID *owner,
        DIDURL *credid, const char **types, size_t typesize, cJSON *json,
        time_t expires, const char *storepass)
{
    Credential *cred = NULL;
    const char *data;
    char signature[SIGNATURE_BYTES * 2];
    int i, rc;
    DIDDocument *doc = NULL;

    assert(issuer && owner && credid);
    assert(types && typesize > 0);
    assert(json);
    assert(expires > 0);
    assert(storepass && *storepass);

    if (!DID_Equals(owner, &credid->did)) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Credential owner is not match with credential did.");
        goto errorExit;
    }

    cred = (Credential*)calloc(1, sizeof(Credential));
    if (!cred) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Malloc buffer for credential failed.");
        goto errorExit;
    }

    if (!DIDURL_Copy(&cred->id, credid))
        goto errorExit;

    //subject
    strcpy(cred->subject.id.idstring, owner->idstring);
    cred->subject.properties = json;

    //set type
    cred->type.size = typesize;
    cred->type.types = (char**)calloc(typesize, sizeof(char*));
    if (!cred->type.types) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Malloc buffer for credential types failed.");
        goto errorExit;
    }
    for (i = 0; i < typesize; i++)
        cred->type.types[i] = strdup(types[i]);

    //set issuer
    DID_Copy(&cred->issuer, &issuer->signer->did);

    //expire and issue date
    cred->expirationDate = expires;
    time(&cred->issuanceDate);

    //proof
    data = Credential_ToJson_ForSign(cred, false, true);
    if (!data)
        goto errorExit;

    rc = DIDDocument_Sign(issuer->signer, &issuer->signkey, storepass, signature,
            1, (unsigned char*)data, strlen(data));
    free((char*)data);
    if (rc)
        goto errorExit;

    strcpy(cred->proof.type, ProofType);
    DIDURL_Copy(&cred->proof.verificationMethod, &issuer->signkey);
    strcpy(cred->proof.signatureValue, signature);
    return cred;

errorExit:
    if (cred)
        Credential_Destroy(cred);
    else
        cJSON_Delete(json);

    if (doc)
        DIDDocument_Destroy(doc);

    return NULL;
}

Credential *Issuer_CreateCredential(Issuer *issuer, DID *owner, DIDURL *credid,
        const char **types, size_t typesize, Property *subject, int size,
        time_t expires, const char *storepass)
{
    cJSON *root, *item;

    if (!issuer ||!owner || !credid || !types || typesize <= 0||
            !subject || size <= 0 || expires <= 0 ||
            !storepass || !*storepass) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    root = cJSON_CreateObject();
    if (!root) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Create property json failed.");
        return NULL;
    }

    for (int i = 0; i < size; i++) {
        item = cJSON_AddStringToObject(root, subject[i].key, subject[i].value);
        if (!item) {
           DIDError_Set(DIDERR_OUT_OF_MEMORY, "Add property failed.");
           cJSON_Delete(root);
           return NULL;
        }
    }

    return issuer_generate_credential(issuer, owner, credid, types, typesize, root,
            expires, storepass);
}

Credential *Issuer_CreateCredentialByString(Issuer *issuer, DID *owner,
        DIDURL *credid, const char **types, size_t typesize, const char *subject,
        time_t expires, const char *storepass)
{
    cJSON *root;

    if (!issuer ||!owner || !credid || !types || typesize <= 0||
            !subject || !*subject || expires <= 0 || !storepass || !*storepass) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    root = cJSON_Parse(subject);
    if (!root) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Deserialize property from json failed.");
        return NULL;
    }

    return issuer_generate_credential(issuer, owner, credid, types, typesize, root,
            expires, storepass);
}

File: DID.Java.SDK/src/main/java/org/elastos/did/Issuer.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did;

import java.io.IOException;
import java.util.Calendar;
import java.util.Date;
import java.util.Map;

import org.elastos.did.exception.DIDStoreException;
import org.elastos.did.exception.InvalidKeyException;
import org.elastos.did.exception.MalformedCredentialException;
import org.elastos.did.exception.MalformedDIDException;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class Issuer {
	private static final String DEFAULT_PUBLICKEY_TYPE = Constants.DEFAULT_PUBLICKEY_TYPE;
	private static final int MAX_VALID_YEARS = Constants.MAX_VALID_YEARS;

	private DIDDocument self;
	private DIDURL signKey;

	public Issuer(DIDDocument doc, DIDURL signKey)
			throws DIDStoreException, InvalidKeyException {
		if (doc == null)
			throw new IllegalArgumentException();

		init(doc, signKey);
	}

	public Issuer(DIDDocument doc)
			throws DIDStoreException, InvalidKeyException {
		this(doc, null);
	}

	public Issuer(DID did, DIDURL signKey, DIDStore store)
			throws DIDStoreException, InvalidKeyException {
		if (did == null || store == null)
			throw new IllegalArgumentException();

		DIDDocument doc = store.loadDid(did);
		if (doc == null)
			throw new DIDStoreException("Can not load DID.");

		init(doc, signKey);
	}

	public Issuer(DID did, DIDStore store)
			throws DIDStoreException, InvalidKeyException {
		this(did, null, store);
	}

	private void init(DIDDocument doc, DIDURL signKey)
			throws DIDStoreException, InvalidKeyException {
		this.self = doc;

		if (signKey == null) {
			signKey = self.getDefaultPublicKey();
		} else {
			if (!self.isAuthenticationKey(signKey))
				throw new InvalidKeyException("Not an authentication key.");
		}

		if (!doc.hasPrivateKey(signKey))
			throw new InvalidKeyException("No private key.");

		this.signKey = signKey;

	}

	public DID getDid() {
		return self.getSubject();
	}

	public DIDURL getSignKey() {
		return signKey;
	}

	public CredentialBuilder issueFor(DID did) {
		if (did == null)
			throw new IllegalArgumentException();

		return new CredentialBuilder(did);
	}

	public CredentialBuilder issueFor(String did) {
		DID _did = null;
		try {
			_did = new DID(did);
		} catch (MalformedDIDException e) {
			throw new IllegalArgumentException(e);
		}

		return issueFor(_did);
	}

	public class CredentialBuilder {
		private DID target;
		private VerifiableCredential credential;

		protected CredentialBuilder(DID target) {
			this.target = target;

			credential = new VerifiableCredential();
			credential.setIssuer(self.getSubject());
		}

		public CredentialBuilder id(DIDURL id) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (id == null)
				throw new IllegalArgumentException();

			credential.setId(id);
			return this;
		}

		public CredentialBuilder id(String id) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			DIDURL _id = id == null ? null : new DIDURL(target, id);
			return id(_id);
		}

		public CredentialBuilder type(String ... types) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (types == null || types.length == 0)
				throw new IllegalArgumentException();

			credential.setType(types);
			return this;
		}

		private Calendar getMaxExpires() {
			Calendar cal = Calendar.getInstance(Constants.UTC);
			if (credential.getIssuanceDate() != null)
				cal.setTime(credential.getIssuanceDate());
			cal.add(Calendar.YEAR, MAX_VALID_YEARS);

			return cal;
		}

		public CredentialBuilder defaultExpirationDate() {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			credential.setExpirationDate(getMaxExpires().getTime());

			return this;
		}

		public CredentialBuilder expirationDate(Date expirationDate) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (expirationDate == null)
				return this;

			Calendar cal = Calendar.getInstance(Constants.UTC);
			cal.setTime(expirationDate);

			Calendar maxExpires = getMaxExpires();
			if (cal.after(maxExpires))
				cal = maxExpires;

			credential.setExpirationDate(cal.getTime());

			return this;
		}

		public CredentialBuilder properties(Map<String, String> properties) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (properties == null || properties.size() == 0)
				throw new IllegalArgumentException();

			ObjectMapper mapper = new ObjectMapper();
			ObjectNode node = mapper.createObjectNode();

			for (Map.Entry<String, String> entry : properties.entrySet())
				node.put(entry.getKey(), entry.getValue());

			return properties(node);
		}

		public CredentialBuilder properties(String json) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (json == null || json.isEmpty())
				throw new IllegalArgumentException();

			ObjectMapper mapper = new ObjectMapper();
			JsonNode node;
			try {
				node = mapper.readTree(json);
			} catch (IOException e) {
				throw new IllegalArgumentException(e);
			}

			return properties(node);
		}

		public CredentialBuilder properties(JsonNode node) {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (node == null || node.size() == 0)
				throw new IllegalArgumentException();

			VerifiableCredential.CredentialSubject subject =
					new VerifiableCredential.CredentialSubject(target);
			subject.setProperties(node);
			credential.setSubject(subject);

			return this;
		}

		public VerifiableCredential seal(String storepass)
				throws MalformedCredentialException, DIDStoreException {
			if (credential == null)
				throw new IllegalStateException("Credential already sealed.");

			if (storepass == null || storepass.isEmpty())
				throw new IllegalArgumentException();

			credential.completeCheck();

			Calendar cal = Calendar.getInstance(Constants.UTC);
			credential.setIssuanceDate(cal.getTime());

			if (!credential.hasExpirationDate())
				defaultExpirationDate();

			String json = credential.toJson(true, true);
			String sig = self.sign(signKey, storepass, json.getBytes());

			VerifiableCredential.Proof proof = new VerifiableCredential.Proof(
					DEFAULT_PUBLICKEY_TYPE, signKey, sig);
			credential.setProof(proof);

			// Invalidate builder
			VerifiableCredential vc = credential;
			this.credential = null;

			return vc;
		}
	}
}
